engine/consensus/dkg/reactor_engine.go

```diff
@@ -0,0 +1,194 @@
+	controller        module.DKGController
+	controllerFactory module.DKGControllerFactory
+	viewEvents        events.Views
+	pollStep          uint64
+}
+
+// NewReactorEngine return a new ReactorEngine.
+func NewReactorEngine(
+	log zerolog.Logger,
+	me module.Local,
+	setups storage.EpochSetups,
+	statuses storage.EpochStatuses,
+	controllerFactory module.DKGControllerFactory,
+	viewEvents events.Views,
+) *ReactorEngine {
+
+	return &ReactorEngine{
+		unit:              engine.NewUnit(),
+		log:               log,
+		me:                me,
+		setups:            setups,
+		statuses:          statuses,
+		controllerFactory: controllerFactory,
+		viewEvents:        viewEvents,
+		pollStep:          DefaultPollStep,
+	}
+}
+
+// Ready implements the module ReadyDoneAware interface. It returns a channel
+// that will close when the engine has successfully
+// started.
+func (e *ReactorEngine) Ready() <-chan struct{} {
+	return e.unit.Ready()
+}
+
+// Done implements the module ReadyDoneAware interface. It returns a channel
+// that will close when the engine has successfully stopped.
+func (e *ReactorEngine) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+// EpochSetupPhaseStarted handles the EpochSetupPhaseStared protocol event. It
+// starts a new controller for the epoch and registers the triggers to regularly
+// query the DKG smart-contract and transition between phases at the specified
+// views.
+func (e *ReactorEngine) EpochSetupPhaseStarted(counter uint64, first *flow.Header) {
+	firstID := first.ID()
+	log := e.log.With().
+		Uint64("counter", counter).
+		Uint64("view", first.View).
+		Hex("block", firstID[:]).
+		Logger()
+	log.Info().Msg("EpochSetup received")
+
+	epochStatus, err := e.statuses.ByBlockID(first.ID())
+	if err != nil {
+		log.Err(err).Msg("could not retrieve epoch state")
+		panic(err)
+	}
+	setup, err := e.setups.ByID(epochStatus.CurrentEpoch.SetupID)
+	if err != nil {
+		log.Err(err).Msg("could not retrieve setup event for current epoch")
+		panic(err)
+	}
+
+	committee := setup.Participants.NodeIDs()
+	index := -1
+	for i, id := range committee {
+		if id == e.me.NodeID() {
+			index = i
+			break
+		}
+	}
+	if index < 0 {
+		err := fmt.Errorf("dkg-processor engine id does not belong to dkg committee")
+		e.log.Err(err).Msg("bad committee")
+		panic(err)
+	}
+
+	controller, err := e.controllerFactory.Create(
+		fmt.Sprintf("dkg-%d", setup.Counter),
+		committee,
+		index,
+		setup.RandomSource,
+	)
+	if err != nil {
+		e.log.Err(err).Msg("could not create DKG controller")
+		panic(err)
+	}
+	e.controller = controller
+
+	e.unit.Launch(func() {
+		e.log.Info().Msg("DKG Run")
+		err := e.controller.Run()
+		if err != nil {
+			e.log.Err(err).Msg("DKG Run error")
+			panic(err)
+		}
+	})
+
+	for view := setup.DKGPhase1FinalView; view > first.View; view -= e.pollStep {
+		e.registerPoll(view)
+	}
+	e.registerPhaseTransition(setup.DKGPhase1FinalView, dkgmod.Phase1, e.controller.EndPhase1)
```

Comment:
Assuming we change the `events.Views` callback semantics as suggested in the other comment, the correctness of this depends on a few things:
* `events.Views` callbacks are invoked in FIFO order
  * otherwise, we may transition to the next dkg phase before submitting the final `Poll` call relative to the final block in the current dkg phase
* all messages retrieved as a result of calling `Poll` are processed before the `Poll` method returns
  * otherwise, we may transition to the next dkg phase before processing the messages from the final `Poll` call relative to the final block in the current dkg phase

I think both these are currently true, but it's not easy to reason that they're true or why it's important for correctness. It would be great to add some comments documenting why these are true, and adding these requirements in the godoc for `events.Views` and `Controller`/`Broker`. It might also help reasoning about the flow if the `Poll` call for the final block of the dkg phase is inline before the phase change call in `registerPhaseTransition`.